### PR TITLE
[mmp] Add stripping of 32-bit dylibs to work with new App Store restrictions

### PR DESCRIPTION
**Pull request by chamons**
- https://github.com/xamarin/xamarin-macios/issues/3367
- Required some test work that might be more obvious by themselves. We can squash when landing.

### CODE REVIEW

File: tools/mmp/driver.cs [hunk 1]
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
...
-
+			
 			bool isStaticLib = real_src.EndsWith (".a", StringComparison.Ordinal);
+			bool isDynamicLib = real_src.EndsWith (".dylib", StringComparison.Ordinal);
+
+			if (isDynamicLib && NativeStrip == NativeStripMode.Strip) {
+				string archToRemove = arch == "i386" ? "x86_64" : "i386";
+				int ret = XcodeRun ("lipo", $"{dest} -remove {archToRemove} -output {dest}");
+				if (ret != 0)
+					throw new MonoMacException (5310, true, "lipo failed with an error code '{0}'. Check build log for details.", ret);
```
Comment (spouliot) [inline tools/mmp/driver.cs]: is this a new error? if so it should be documented

File: tools/mmp/driver.cs [hunk 2]
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
...
+				if (name != "MonoPosixHelper")
+					ErrorHelper.Warning (1501, $"{name} was stripped of architecture {archToRemove} to comply with App Store restrictions. This could break exisiting codesigning signatures. Consider stripping the library with lipo or disabling with --native-strip=skip");
```
Comment (spouliot) [inline tools/mmp/driver.cs]: same

File: tools/mmp/driver.cs [hunk 3]
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
 				// install_name_tool gets angry if you copy in a read only native library
 				CopyFileAndRemoveReadOnly (real_src, dest);
 			}
-
+			
 			bool isStaticLib = real_src.EndsWith (".a", StringComparison.Ordinal);
+			bool isDynamicLib = real_src.EndsWith (".dylib", StringComparison.Ordinal);
```
Comment (spouliot) [inline tools/mmp/driver.cs]: what happens for frameworks ?

File: tests/mmptest/src/CodeStrippingTests.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+			MMPTests.RunMMPTest (tmpDir =>
+			{
+				 TI.UnifiedTestConfig test = CreateStripTestConfig (nativeStrip, tmpDir);
+
+				 string buildOutput = TI.TestUnifiedExecutable (test).BuildOutput;
+				 Assert.AreEqual (debugStrips, didStrip (buildOutput), "Debug lipo usage did not match expectations");
+
+				 test.Release = true;
+				 buildOutput = TI.TestUnifiedExecutable (test).BuildOutput;
+				 Assert.AreEqual (releaseStrips, didStrip (buildOutput), "Release lipo usage did not match expectations");
```
Comment (rolfbjarne) [inline tests/mmptest/src/CodeStrippingTests.cs]: Maybe also test that the binary only has the desired architectures? We already have code to get the architectures of a binary file: https://github.com/xamarin/xamarin-macios/blob/bec116c273ab119c648522a28c3f650881ed4d9e/tests/mtouch/MTouch.cs#L3996

File: tests/common/mac/ProjectTestHelpers.cs [hunk 5]
```diff
@@ -145,11 +152,15 @@ public static string BuildProject (string csprojTarget, bool isUnified, bool dia
 			if (isUnified) {
 				buildArgs.Append (diagnosticMSBuild ? " /verbosity:diagnostic " : " /verbosity:normal ");
 				buildArgs.Append (" /property:XamarinMacFrameworkRoot=" + rootDirectory + "/Library/Frameworks/Xamarin.Mac.framework/Versions/Current ");
-
-				if (!string.IsNullOrEmpty (configuration))
-					buildArgs.Append ($" /property:Configuration={configuration} ");
-			} else
+			}
```
Comment (rolfbjarne) [inline tests/common/mac/ProjectTestHelpers.cs]: Style: `else` goes on the same line as the closing brace.

File: tools/mmp/driver.cs [hunk 6]
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
 				// install_name_tool gets angry if you copy in a read only native library
 				CopyFileAndRemoveReadOnly (real_src, dest);
 			}
-
+
```
Comment (rolfbjarne) [inline tools/mmp/driver.cs]: Whitespace noise.

File: tools/mmp/driver.cs (hunk 1 above)
Comment (chamons) [reply tools/mmp/driver.cs]: Yes, I forgot.

File: tools/mmp/driver.cs (hunk 3 above)
Comment (chamons) [reply tools/mmp/driver.cs]: Nothing, but something very likely should. Let me get someone in to test this case in submission to Apple.

**Comment by chamons** [comment]
I believe I've handled the current code review issues and added framework stripping.

I did not build after merging master, and a mono bump took too long before end of day. I'll check back tomorrow. Please consider re-reviewing as a lot has changed.

File: tools/mmp/driver.cs [hunk 7]
```diff
@@ -366,6 +374,22 @@ static void Main2 (string [] args)
...
+					case "strip":
+						NativeStrip = NativeStripMode.Strip;
+						break;
+					case "skip":
+						NativeStrip = NativeStripMode.Skip;
+						break;
+					default:
+						ErrorHelper.Error (26, $"Could not parse the command line argument '-native-strip:{v}'");
+						break;
+					}
```
Comment (spouliot) [inline tools/mmp/driver.cs]: I wonder if that should be added in the new optimizations ? c.c. @rolfbjarne 
It behaves just like them, i.e. a sane *default* (computed based on the project settings) and overrides for *enabling* and *disabling* it.
In any case it needs to be documented outside the errors code it introduce.

File: tools/mmp/driver.cs [hunk 8]
```diff
@@ -1720,6 +1755,17 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
...
 
+		static void LipoLibrary (string name, string dest)
+		{
+			string archToRemove = arch == "i386" ? "x86_64" : "i386";
+
+			int ret = XcodeRun ("lipo", $"{dest} -remove {archToRemove} -output {dest}");
+			if (ret != 0)
+				throw new MonoMacException (5311, true, "lipo failed with an error code '{0}'. Check build log for details.", ret);
+			if (name != "MonoPosixHelper")
+				ErrorHelper.Warning (2108, $"{name} was stripped of architecture {archToRemove} to comply with App Store restrictions. This could break exisiting codesigning signatures. Consider stripping the library with lipo or disabling with --native-strip=skip");
```
[the remark this answers is not in the document]
Comment (chamons) [reply tools/mmp/driver.cs]: The "obvious" thing to do here is to first write a test that says "if I'm 64-bit and i have a 64-bit library there should be no warning".

And yes, it looks like that test would fail right now. :(

**Comment by chamons** [comment]
I've moved to --optimize=-trim-architectures instead of a new option and fixed the location of common string utils.

I'm going to work on fixing the likely every present warning w\ native libraries.

@rolfbjarne please review my optimization changes, first time I've added one.

File: tools/common/Driver.cs [hunk 9]
```diff
@@ -142,7 +142,12 @@ static void AddSharedOptions (Application app, Mono.Options.OptionSet options)
 					"    blockliteral-setupblock: By default enabled when using the static registrar. Optimizes calls to BlockLiteral.SetupBlock to avoid having to calculate the block signature at runtime.\n" +
 					"    inline-intptr-size: By default enabled for builds that target a single architecture (requires the linker). Inlines calls to IntPtr.Size to load a constant value. Makes the app smaller, and slightly faster at runtime.\n" +
 					"    inline-dynamic-registration-supported: By default always enabled (requires the linker). Optimizes calls to Runtime.DynamicRegistrationSupported to be a constant value. Makes the app smaller, and slightly faster at runtime.\n" +
-					"    register-protocols: Remove unneeded metadata for protocol support. Makes the app smaller and reduces memory requirements.\n",
+					"    register-protocols: Remove unneeded metadata for protocol support. Makes the app smaller and reduces memory requirements.\n" +
+#if !MONOTOUCH
+					"    trim-architectures: Remove unneeded architectures from bundled native libraries. Makes the app smaller and necessary for macOS App Store submissions.\n",
```
[the remark this answers is not in the document]
Comment (spouliot) [reply tools/common/Driver.cs]: **required** ? more than necessary ?

File: docs/website/mmp-errors.md [hunk 10]
```diff
@@ -248,6 +250,15 @@ To remove the warning either remove the optimization argument to mmp, or pass
 By default this option will be automatically enabled whenever it's possible
 and safe to do so.
 
+### <a name="MM2108"/>MM2108: '{0}' was stripped of architecture '{1}' to comply with App Store restrictions. This could break exisiting codesigning signatures. Consider stripping the library with lipo or disabling with --optimize=-trim-architectures");
+
+The App Store now rejects applications which contain libraries and frameworks which contain 32-bit variants. The library was stripped of unused archtectures when copied into the final application bundle.
+
+This is in general safe, and will reduce application size as an added benefit, but may break existing codesigning signatures if it was already signed.
```
[the remark this answers is not in the document]
Comment (chamons) [reply docs/website/mmp-errors.md]: We do, but if there was some pre-existing signature on the frameowork we're copying we'd be breaking that and resigning over top.

Maybe it isn't worth mentioning that corner case. In 99% of use cases we're going to fix things when signing at the end.

**Comment by chamons** [comment]
Ok, I added ThirdPartyLibrary_WithCorrectBitness_ShouldNotStripOrWarn test and fixed the others so they would fail without my fix.

I could make the logic more complicated in mmp but bailing on < 2 arch in a lib works and seemed simpler. Thoughts?

File: docs/website/mmp-errors.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply docs/website/mmp-errors.md]: My point is that if we're always signing dylibs and frameworks, stripping unused architectures from them won't change anything (since we always overwrite any existing signature anyway).

[the remark this answers is not in the document]
Comment (rolfbjarne) [reply docs/website/mmp-errors.md]: Ah, I see, it's not required to sign the .app like it is for XI, in which case this does make sense. So nvm.

**Comment by chamons** [comment]
Some test failures are legit and need looking at.

**Comment by chamons** [comment]
I can not reproduce them currently with ```make run-mac-unified-mmptest``` or mmptest.csproj. Trying another build in parallel.

**Comment by chamons** [comment]
build

**Comment by chamons** [comment]
I can reproduce them when run from the command line but not from the IDE, which is strange. In any case, looking.

**Comment by chamons** [comment]
This PR has started to turn into a monster, just like every other time I touch mmp...

A few more tests are failing locally after the last set of pushes. Looking now.

**Comment by chamons** [comment]
Looks clean enough locally now. 🤞

**Comment by chamons** [comment]
Looks like some tests are comparing strings that changed when I added a message. Fixing.

**Comment by chamons** [comment]
Some known test failures:

https://github.com/xamarin/maccore/issues/640
https://github.com/xamarin/maccore/issues/656
https://github.com/xamarin/maccore/issues/629

Looking into protocol failures, feel unrelated.

**Comment by chamons** [comment]
https://github.com/xamarin/maccore/issues/655 for the protocol failures.
